### PR TITLE
Added callback condition support

### PR DESCRIPTION
**Pull request by wodCZ**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -231,9 +233,98 @@ $grid->addFilter('name', Condition::LIKE, [
...
 
-Now you only have to display form inputs in your template.
+### Rendering filter inputs
+
+You have to render filter inputs yourself. Your template may then look like this:
+
+**Please note that you do not have to render form tags as NoGrid will do it for you.**
+
+```smarty
+<table n:no-grid="personsGrid" class="table table-striped">
```
Comment (davidkudera) [inline README.md]: `peopleGrid`? :trollface:

File: README.md [hunk 2]
```diff
@@ -231,9 +233,98 @@ $grid->addFilter('name', Condition::LIKE, [
...
+	<thead>
+	<tr>
```
Comment (davidkudera) [inline README.md]: Indentation (I know, PhpStorm...)

File: README.md [hunk 3]
```diff
@@ -231,9 +233,98 @@ $grid->addFilter('name', Condition::LIKE, [
...
+			Name
+			<br>
+			<input n:name="name">
+			<input n:name="search">
+		</th>
+	</tr>
+	</thead>
+
+	<tbody>
+	<tr n:no-grid-data-as="$line">
```
Comment (davidkudera) [inline README.md]: Indentation

File: README.md [hunk 4]
```diff
@@ -231,9 +233,98 @@ $grid->addFilter('name', Condition::LIKE, [
...
+$form = new \Nette\Application\UI\Form;
+
+$form->addText('fullname');
+$form->addSubmit('search', 'Search');
+
+$grid->setFilteringForm($form);
+
+$grid->addFilter('person', Condition::CALLBACK, [], function (QueryBuilder $queryBuilder, $value) {
+	$queryBuilder->andWhere("CONCAT(p.givenName,' ',p.familyName) LIKE :fullname")
+		 		 ->setParameter('fullname', "%{$value}%");
```
Comment (davidkudera) [inline README.md]: Indentation

File: src/Condition.php [hunk 5]
```diff
@@ -46,6 +47,10 @@ public function __construct($column, $value, $type = self::SAME, array $options
 		$this->value = $value;
 		$this->type = $type;
 		$this->options = $options;
+
+		if($type === self::CALLBACK && ( ! array_key_exists(self::CALLBACK, $options) || ! is_callable($options[self::CALLBACK]))){
```
Comment (davidkudera) [inline src/Condition.php]: Please, can you remove the whitespaces like this: `(!array_key_exists(self::CALLBACK, $options) || !is_callable($options[self::CALLBACK]))`?

File: src/DataSource/ArrayDataSource.php [hunk 6]
```diff
@@ -68,13 +69,24 @@ public function fetchData()
 	public function filter(array $conditions)
 	{
 		foreach ($conditions as $condition) {
-			$this->data = array_filter($this->fetchData(), function($row) use ($condition) {
-				if (!isset($row[$condition->getColumn()])) {
-					return false;
+			if($condition->getType() === Condition::CALLBACK){
```
Comment (davidkudera) [inline src/DataSource/ArrayDataSource.php]: and put spaces before the curly braces?

File: src/exceptions.php [hunk 7]
```diff
@@ -29,3 +29,10 @@ class InvalidArgumentException extends LogicException {}
  * @author David Kudera <kudera.d@gmail.com>
  */
 class NotImplementedException extends LogicException {}
+
+
+/**
+ *
+ * @author Martin Janeček <admin@ikw.cz>
+ */
+class InvalidStateException extends \RuntimeException {}
```
Comment (davidkudera) [inline src/exceptions.php]: Could you please create new `RuntimeException` class and extend `InvalidStateException` from the new one?

**Comment by davidkudera** [comment]
that's good :+1: 

unfortunately I found one problem, but not in your code. The reason I don't like for example `o5/grido` is, that many times I had parts of DQL in my components. But I wanted to use just my QueryObjects, because they should contain all the logic for creating queries. 

So it would be nice I think if we are able to use something like this:

``` php
$grid->addFilter('fullname', Condition::CALLBACK, [], function (QueryObject $query, $value) { 
    $query->byFullname($value);
});
```

So then arguments should be the like this I think:
- `ArrayDataSource`: `array` of all items
- `Doctrine\DataSource`: `QueryBuilder` instance
- `Doctrine\QueryObjectDataSource`: `QueryObject` instance
- `Doctrine\QueryFunctionDataSource`: `AbstractQuery` instance

Downside to this is that it would complicate things a bit. Also those callback filters will need to be called before any other filters. Probably before this line: https://github.com/wodCZ/Nette-NoGrid/blob/master/src/DataSource/Doctrine/DoctrineDataSource.php#L70 . Of course also for other  datasource's too.

What do you think about it?

**Comment by wodCZ** [comment]
Arguments should be like you say I think, but I'm not sure about implementation.
I'm not near NoGrid in my project currently, so I'll have to look at that in free time, probably/hopefully this weekend.

I'd like to not separate callback filters from other filters while I understand the need for calling them before any other filters.

Probably sorting conditions would be enough?

Then we could call them same way and only change needed would be to provide overridable method [here](https://github.com/wodCZ/Nette-NoGrid/blob/master/src/DataSource/Doctrine/BaseDataSource.php#L193), something like `static::callConditionCallback($notSureAboutParameters)`.

**Comment by davidkudera** [comment]
I also want to keep all conditions together if possible. If you'll have any time, it would be really greate :wink:

**Pull request closed without merging**